### PR TITLE
[DEV-707] Drop unused indexes

### PR DESCRIPTION
**Pull request by vmarkovtsev**

### CODE REVIEW

**Comment by vmarkovtsev** [comment]
@dennwc refer to this file for the explanation:
[indexes.xlsx](https://github.com/athenianco/metadata/files/5096201/indexes.xlsx)

**Review by dennwc — CHANGES_REQUESTED** [review]
@vmarkovtsev Please don't forget that MD also uses some of the indexes.

For example, `_rev_cover` indexes on edges are used by the consistency checker. `created_at` are used for debug sometimes, they help list recent PR events.

Some of them are also sane and general-purpose, e.g. `bot_login_cover`, `commit_sha_cover`, etc.

So the cleanup requires more consideration.

**Comment by vmarkovtsev** [comment]
I collected the spreadsheet by querying the global index usage stats on production. All those are not really used.

**Comment by vmarkovtsev** [comment]
Also @dennwc when you debug, you don't really need super-fast responses. All those `created_at`-s can be safely removed IMO.

**Comment by vmarkovtsev** [comment]
Also WDYM general-purpose? There is no other metadata user except the API. If something is not used, it should be removed.

**Comment by vmarkovtsev** [comment]
@se7entyse7en What do you think about this?

**Comment by vmarkovtsev** [comment]
@dennwc Also please note how those rev_cover indexes are deleted: after they are created so that when we see that something starts working slow, we delete the corresponding lines.

**Comment by dennwc** [comment]
@vmarkovtsev Yes, I'm aware of the document. That's a good way of checking it. Bu the stats may change over time. We recently dropped the prod instance, it might have influenced the stats as well.

If `rev` indexes are not used, it only means that PG prefers to scan over the whole table when checking consistency. This is temporary, however because as soon as we isolate data it will be able to scan those indexes properly.

`created_at` May still be useful. Sure, we don't need it to be "super fast", but slowing the instance by scanning _all_ events is  unreasonable IMO.

By general-purpose I mean that querying on `login` is reasonable and will happen. Ideally, the `login` must be `UNIQUE` across users, bots and orgs. However, due to eventual consistency in MD it's not enforced. The query is still valid though, and must be indexed. API used logins for quite some time as well.

**Comment by vmarkovtsev** [comment]
> but slowing the instance by scanning all events is unreasonable IMO
You complain that it is getting slow to update the tables, I propose a good solution to remove 100500 unnecessary indexes, now you are fighting to leave some debug stuff which is not really needed (sorry, but this is my opinion). Well, I am not directly affected, so I let you step on this rake :bomb: 

> The query is still valid though, and must be indexed. API used logins for quite some time as well.
Let's not argue with the API author about which queries require which indexes :smile:  If I see from my *extensive* tracing that there is an index that must be added, I will add it. Otherwise, I want to delete it without regrets. In this case, the logins are not needed. YAGNI and KISS on the DB level.

**Comment by dennwc** [comment]
> You complain that it is getting slow to update the tables

I complained about specific tables, e.g. `commits`, which are _huge_. It's not the same for _every_ table in the DB. Please collect DB row count stats and pick the first few tables instead of wiping every index you don't use.

> Let's not argue with the API author about which queries require which indexes.

Are you trying to make it personal or something? That's not a constructive behavior :stuck_out_tongue: 

What I'm saying is that there are the second system (MD itself) that has it's own needs. Please respect it the same way as you respect your own system. And as a maintainer of MD I say that you shouldn't wipe these indexes.

**Comment by vmarkovtsev** [comment]
Sorry for being an annoying pain in the ass, but can you demonstrate the place in MD code where you query `login`?

**Comment by vmarkovtsev** [comment]
> Are you trying to make it personal or something?

Not at all, it just looked as if you speculated against the real API requirements.

**Comment by se7entyse7en** [comment]
> @dennwc refer to this file for the explanation:
indexes.xlsx

@vmarkovtsev if next time you could use Google spreadsheet that would make it easier to read 🙏 , just put it inside the `Engineering` folder: 
https://docs.google.com/spreadsheets/d/1dTu9zRVOvSwaGAsrmIJwaWUaKTY4HLzedAb9kGdZ8l0/edit

Do the colors have some meaning? Or Google spreadsheet just messed up the import?

> @vmarkovtsev Yes, I'm aware of the document. That's a good way of checking it. Bu the stats may change over time. We recently dropped the prod instance, it might have influenced the stats as well.

This is the first thing that I thought. I'm not a PG expert, but since we dropped it recently it's true that stats could not reflect much the real usage of them. Probably it reflects the current API usage since it performs much more reads compared to other clients.

> You complain that it is getting slow to update the tables, I propose a good solution to remove 100500 unnecessary indexes, now you are fighting to leave some debug stuff which is not really needed (sorry, but this is my opinion). Well, I am not directly affected, so I let you step on this rake 💣

I don't know exactly how debugging usually works on MD (tracking inconsistencies, missing events, etc.) and whether it may involve some "common queries". It's true that super-fast response are not needed since those are not facing the customers, but if there are some queries that are frequently used for debugging require some indexes I'd rather keep them as also @dennwc's time has a cost and I don't want him to spend time waiting for query responses. The same reasoning applies to you @vmarkovtsev in case you also need to debug things and you end up running some "common queries" (maybe for traversing the commits DAG for example to investigate releases).

> I complained about specific tables, e.g. commits, which are huge. It's not the same for every table in the DB. Please collect DB row count stats and pick the first few tables instead of wiping every index you don't use.

I'm wondering whether it makes sense to also consider the creation timestamp of an index, as of course oldest index have been used more for sure 😄.

> Are you trying to make it personal or something? That's not a constructive behavior 😛

hahahaha 😂 
I agree that making things personal is not a constructive behaviour, but I don't think that's what @vmarkovtsev meant. Of course @vmarkovtsev knows API's needs and those needs are reflected by the stats he collected and as @dennwc said there are also other clients of that DB with their needs and as mentioned before some of them could be useful for debugging.

To recap, gathering the stats of the indexes is a good idea so that we could eventually reclaim some space and maybe also improve some perf due to the overhead of updating all of them. Unfortunately, afaik we cannot quantify this nor whether during some queries the reads are slowed-down by some index updates. I guess that among all those unused indexes there's only a subset that @dennwc thinks that could be useful for either debugging or other MD clients. I don't think that this is critical for now, but @dennwc could go over those that are marked as unused indexes according to the current stats and filter those he agreed are not useful and that could be deleted. Then in the following weeks we could gather these stats again.

**Comment by dennwc** [comment]
So to recap:
- We must keep `_rev_` indexes, since MD will use them soon.
- We should keep the `login` and `created_at` since it helps with debugging and are not that impactful.
- Everything else can be optimized and cleaned up.
- This PR needs a rebase (basically all SQL is now in `./sql`).

**Comment by vmarkovtsev** [comment]
Sorry fellows, I have to abandon this work as I have higher priority tasks to do now.
